### DISCUSSION TITLE
Compiling the bootloader

### DISCUSSION THREAD

**Discussion by VasiliuVlad** [SpenceKonde/DxCore#179] [Q&A]
Hello,
I was looking into the bootloader trying to compile it  from code.
I have several questions here (maybe some of them are dumb .. so I apologize in advance :) )

1. Makefile : I am really confused about this part (from line 139)
```
optiboot_%.hex: optiboot_%.elf
	$(OBJCOPY) -j .text -j .data -j .version -j .spmtarg --set-section-flags .version=alloc,load --set-section-flags .spmtarg=alloc,load -O ihex $< $@

optiboot_%.elf:	optiboot_x.c FORCE
	$(CC) $(CFLAGS) $(CPU_OPTIONS) $(LED_OPTIONS) $(UART_OPTIONS) $(POR_CMD) $(COMMON_OPTIONS) $(LDFLAGS) $(PACK_OPT) -mmcu=$(TARGET) -o $@ $<
	$(SIZE) $@
	$(LISTING) $@ > optiboot_$*.lst
```
As far as I can understand the file which is compiled/linked = built  is optiboot_x.c   but optiboot_**d**x.c is provided .
Am I getting this wrong ? 
2. The target is somehow not recognized by the avr-gcc .  Do you use atmel studio? is there a  gnu gcc version that I can use ?

Thanks in advance,
Best regards,
Vlad

**Comment by VasiliuVlad** [comment]
And an additional thing ... 
```
      // read a page worth of contents
      bufPtr = buff.bptr;
      do {
        *bufPtr++ = getch();
      } while (--length);

      // Read command terminator, start reply
      verifySpace();
      writebuffer(desttype, buff, address, savelength);
```
This part is commented out .... but it seems that this is exactly the part which writes the Flash .

Am I getting this wrong ?

**Comment by SpenceKonde** [comment]
> Hello, I was looking into the bootloader trying to compile it from code. I have several questions here (maybe some of them are dumb .. so I apologize in advance :) )
> 
> 1. Makefile : I am really confused about this part (from line 139)
> 
> ```
> optiboot_%.hex: optiboot_%.elf
> 	$(OBJCOPY) -j .text -j .data -j .version -j .spmtarg --set-section-flags .version=alloc,load --set-section-flags .spmtarg=alloc,load -O ihex $< $@
> 
> optiboot_%.elf:	optiboot_x.c FORCE
> 	$(CC) $(CFLAGS) $(CPU_OPTIONS) $(LED_OPTIONS) $(UART_OPTIONS) $(POR_CMD) $(COMMON_OPTIONS) $(LDFLAGS) $(PACK_OPT) -mmcu=$(TARGET) -o $@ $<
> 	$(SIZE) $@
> 	$(LISTING) $@ > optiboot_$*.lst
> ```
> 
> As far as I can understand the file which is compiled/linked = built is optiboot_x.c but optiboot_**d**x.c is provided . Am I getting this wrong ? 2. The target is somehow not recognized by the avr-gcc . Do you use atmel studio? is there a gnu gcc version that I can use?

1. That's a default, but all the files we distribute were built with a part definition below that line. Probably should be dx - though it is not expected to work when built like that. 
2. Built with the azduino tool chain that the core uses (see json file  for url for your platform). I start with arduino 1.0.6 .zip package (for stuff like make) on windows, copied the latest tool chain on top of the existing one in hardware/tools to get the new device-spec and io headers), copied optiboot_dx into hardware/bootloaders and then deleted everything outside of hardware/tools and hardware/bootloaders. I run the batch file from command prompt in that scenario, and it works. 

In other words, I didn't set it up, and don't fully understand it or makefiles in general. I don't like and am not good at that crap ("I write code, not makefiles". I really don't use anything more than sublime text and the makefile. Which I took from megaTinyCore, as you can see by the filename (optiboot_x over there too, I guess that was chosen... I guess because they initially appeared to be xmega-tinies? It was fairly common outside arduino at first - before it became clear that they were the vanguard of the modern avr product line. I'd have called it optiboot_xt since it's the AVRxt instruction set, I suppose.) I know how to modify the makefile to do what I want, but I couldn't write it from scratch.

**Comment by SpenceKonde** [comment]
And yeah the comment before that block wasn't closed properly,  I added that after the latest serialupdi improvements, bootloader has not been compiled since before I was doing that - as I learn new things that make me realize that my old comments or design decisions were wrong, before I get a chance to correct them I add comments regarding it to the code so it will be there next time I look and I will remember why I thought there was something I needed to do there.
